aws/s3.go

```diff
@@ -304,119 +304,181 @@ func excludeBucketByREList(bucketName string, reList []*regexp.Regexp) bool {
+			Key: obj.Key,
+		})
 	}
+	_, err := svc.DeleteObjects(
+		&s3.DeleteObjectsInput{
+			Bucket: bucketName,
+			Delete: &s3.Delete{
+				Objects: objectIdentifiers,
+				Quiet:   aws.Bool(false),
+			},
+		},
+	)
+	return err
+}
 
-	logging.Logger.Infof("Deleting - Bucket: %s - objects: %d", *bucketName, totalObjects)
-
-	totalBatches := totalObjects / batchSize
-	if totalObjects%batchSize != 0 {
-		totalBatches = totalBatches + 1
+// deleteObjectVersions will delete the provided object versions from the specified bucket.
+func deleteObjectVersions(svc *s3.S3, bucketName *string, objectVersions []*s3.ObjectVersion) error {
+	if len(objectVersions) == 0 {
+		logging.Logger.Debugf("No object versions returned in page")
+		return nil
 	}
-	batchCount := 1
 
-	// Batch the delete operation
-	for i := 0; i < len(objects); i += batchSize {
-		j := i + batchSize
-		if j > len(objects) {
-			j = len(objects)
-		}
+	objectIdentifiers := []*s3.ObjectIdentifier{}
+	for _, obj := range objectVersions {
+		objectIdentifiers = append(objectIdentifiers, &s3.ObjectIdentifier{
+			Key:       obj.Key,
+			VersionId: obj.VersionId,
+		})
+	}
+	_, err := svc.DeleteObjects(
+		&s3.DeleteObjectsInput{
+			Bucket: bucketName,
+			Delete: &s3.Delete{
+				Objects: objectIdentifiers,
+				Quiet:   aws.Bool(false),
+			},
+		},
+	)
+	return err
+}
 
-		logging.Logger.Debugf("Deleting - %d-%d objects of batch %d/%d - Bucket: %s", i+1, j, batchCount, totalBatches, *bucketName)
+// deleteDeletionMarkers will delete the provided deletion markers from the specified bucket.
+func deleteDeletionMarkers(svc *s3.S3, bucketName *string, objectDelMarkers []*s3.DeleteMarkerEntry) error {
+	if len(objectDelMarkers) == 0 {
+		logging.Logger.Debugf("No deletion markers returned in page")
+		return nil
+	}
 
-		delObjects := objects[i:j]
-		_, err = svc.DeleteObjects(
-			&s3.DeleteObjectsInput{
-				Bucket: bucketName,
-				Delete: &s3.Delete{
-					Objects: delObjects,
-					Quiet:   aws.Bool(false),
-				},
+	objectIdentifiers := []*s3.ObjectIdentifier{}
+	for _, obj := range objectDelMarkers {
+		objectIdentifiers = append(objectIdentifiers, &s3.ObjectIdentifier{
+			Key:       obj.Key,
+			VersionId: obj.VersionId,
+		})
+	}
+	_, err := svc.DeleteObjects(
+		&s3.DeleteObjectsInput{
+			Bucket: bucketName,
+			Delete: &s3.Delete{
+				Objects: objectIdentifiers,
+				Quiet:   aws.Bool(false),
 			},
-		)
-		if err != nil {
-			return err
-		}
+		},
+	)
+	return err
+}
+
+// nukeAllS3BucketObjects batch deletes all objects in an S3 bucket
+func nukeAllS3BucketObjects(svc *s3.S3, bucketName *string, batchSize int) error {
+	versioningResult, err := svc.GetBucketVersioning(&s3.GetBucketVersioningInput{
+		Bucket: bucketName,
+	})
+	if err != nil {
+		return err
+	}
 
-		logging.Logger.Infof("[OK] - %d-%d objects of batch %d/%d - Bucket: %s - deleted", i+1, j, batchCount, totalBatches, *bucketName)
+	isVersioned := versioningResult.Status != nil && *versioningResult.Status == "Enabled"
```

Comment:
NIT: I know the code was this way before, but a minor improvement:

```suggestion
	isVersioned := aws.StringValue(versioningResult.Status) == "Enabled"
```